src/zon.zig

```diff
@@ -0,0 +1,831 @@
+    for (field_values) |*v| v.* = .none;
+
+    for (0..fields.names.len) |i| {
+        const field_name = try ip.getOrPutString(
+            gpa,
+            self.sema.pt.tid,
+            fields.names[i].get(self.file.zoir.?),
+            .no_embedded_nulls,
+        );
+        const field_node = fields.vals.at(@intCast(i));
+
+        const name_index = struct_info.nameIndex(ip, field_name) orelse {
+            return self.fail(
+                .{ .node_abs = field_node.getAstNode(self.file.zoir.?) },
+                "unexpected field '{}'",
+                .{field_name.fmt(ip)},
+            );
+        };
+
+        const field_type = Type.fromInterned(struct_info.field_types.get(ip)[name_index]);
+        if (field_values[name_index] != .none) {
+            const field_node_ast = field_node.getAstNode(self.file.zoir.?);
+            const field_name_token = self.file.tree.firstToken(field_node_ast) - 2;
+            return self.fail(
+                .{ .token_abs = field_name_token },
+                "duplicate field '{}'",
+                .{field_name.fmt(ip)},
+            );
+        }
+
+        field_values[name_index] = try self.lowerExpr(field_node, field_type);
+
+        if (struct_info.comptime_bits.getBit(ip, name_index)) {
+            const val = ip.indexToKey(field_values[name_index]);
+            const default = ip.indexToKey(field_defaults[name_index]);
+            if (!val.eql(default, ip)) {
+                const field_node_ast = field_node.getAstNode(self.file.zoir.?);
+                const field_name_token = self.file.tree.firstToken(field_node_ast) - 2;
+                return self.fail(
+                    .{ .token_abs = field_name_token },
+                    "value stored in comptime field does not match the default value of the field",
+                    .{},
+                );
+            }
+        }
+    }
+
+    for (0..field_values.len) |i| {
+        if (field_values[i] == .none and i < field_defaults.len) {
+            field_values[i] = field_defaults[i];
+        }
+    }
+
+    const field_names = struct_info.field_names.get(ip);
+    for (field_values, field_names) |*value, name| {
+        if (value.* == .none) return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "missing field {}",
+            .{name.fmt(ip)},
+        );
+    }
+
+    return self.sema.pt.intern(.{ .aggregate = .{ .ty = res_ty.toIntern(), .storage = .{
+        .elems = field_values,
+    } } });
+}
+
+fn lowerPointer(self: LowerZon, node: Zoir.Node.Index, res_ty: Type) !InternPool.Index {
+    const ip = &self.sema.pt.zcu.intern_pool;
+    const gpa = self.sema.gpa;
+
+    const ptr_info = res_ty.ptrInfo(self.sema.pt.zcu);
+
+    if (ptr_info.flags.size != .Slice) {
+        return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "non slice pointers are not available in ZON",
```

Comment:
Okay, this is an interesting question -- should we allow implicitly lowering ZON through single-item pointers (i.e. when we expect a `*u32`, allow `123` and treat it as `&123`)? I don't really see a strong reason to disallow it, but at the same time can't see any real use cases. However, it might be more consistent, because this is basically already what we're allowing with slices.

Reply:
We discussed this offline and found some reasonable use cases: recursive types like config files that include trees, types where you want to only allocate bulky parts if they have data.

I've now implemented support for this.